File: packages/dtj-go/e2e_test.go
package dtj

import (
	"os"
	"path/filepath"
	"testing"
)

// TestE2E is an opt-in end-to-end test that requires a real dtj-agent binary
// and Unix socket support. It is skipped unless DTJ_RUN_AGENT_E2E=1 is set.
func TestE2E(t *testing.T) {
	if os.Getenv("DTJ_RUN_AGENT_E2E") != "1" {
		t.Skip("E2E test skipped. Set DTJ_RUN_AGENT_E2E=1 to run.")
	}

	// Create a temporary directory for traces
	tmpDir := t.TempDir()

	trace := Open(Config{
		ProducerName:    "e2e-test",
		ProducerVersion: "0.1.0",
		DataDir:         tmpDir,
	})
	if trace == nil {
		t.Fatal("Open returned nil")
	}
	defer trace.Close()

	// Emit various event types
	tests := []Event{
		{
			Domain:      "api",
			Category:    "request",
			Name:        "completed",
			Severity:    SeverityInfo,
			FieldName:   "duration_ms",
			Value:       12.5,
			Correlation: "request-42",
		},
		{
			Domain:    "test",
			Category:  "bool",
			Name:      "flag",
			Severity:  SeverityInfo,
			FieldName: "enabled",
			Value:     true,
		},
		{
			Domain:    "test",
			Category:  "int",
			Name:      "count",
			Severity:  SeverityInfo,
			FieldName: "value",
			Value:     42,
		},
		{
			Domain:    "test",
			Category:  "bigint",
			Name:      "large",
			Severity:  SeverityInfo,
			FieldName: "value",
			Value:     int64(123456789012345),
		},
		{
			Domain:    "test",
			Category:  "float",
			Name:      "ratio",
			Severity:  SeverityInfo,
			FieldName: "value",
			Value:     3.14159,
		},
		{
			Domain:    "test",
			Category:  "bytes",
			Name:      "payload",
			Severity:  SeverityInfo,
			FieldName: "data",
			Value:     []byte{0xDE, 0xAD, 0xBE, 0xEF},
		},
	}

	for _, event := range tests {
		if err := trace.Emit(event); err != nil {
			t.Fatalf("Emit failed: %v", err)
		}
	}

	// Close the session
	if err := trace.Close(); err != nil {
		t.Fatalf("Close failed: %v", err)
	}

	// Verify .dtj file was created
	files, err := filepath.Glob(filepath.Join(tmpDir, "*.dtj"))
	if err != nil {
		t.Fatalf("failed to glob dtj files: %v", err)
	}
	if len(files) == 0 {
		t.Fatal("no .dtj file created")
	}

	// Verify file has content and correct magic bytes
	for _, file := range files {
		content, err := os.ReadFile(file)
		if err != nil {
			t.Fatalf("failed to read file: %v", err)
		}
		if len(content) == 0 {
			t.Fatal("session file is empty")
		}
		// DTJ v1 magic: "DTJ\1" (0x44 0x54 0x4A 0x01)
		if len(content) < 4 || content[0] != 0x44 || content[1] != 0x54 || content[2] != 0x4A || content[3] != 0x01 {
			t.Fatalf("invalid DTJ magic bytes: %v", content[:4])
		}
	}
}

// TestE2ESeverityLevels tests all severity levels
func TestE2ESeverityLevels(t *testing.T) {
	if os.Getenv("DTJ_RUN_AGENT_E2E") != "1" {
		t.Skip("E2E test skipped. Set DTJ_RUN_AGENT_E2E=1 to run.")
	}

	tmpDir := t.TempDir()

	trace := Open(Config{
		ProducerName:    "e2e-test",
		ProducerVersion: "0.1.0",
		DataDir:         tmpDir,
	})
	if trace == nil {
		t.Fatal("Open returned nil")
	}
	defer trace.Close()

	severities := []Severity{SeverityDebug, SeverityInfo, SeverityWarn, SeverityError, SeverityFatal}
	for _, sev := range severities {
		if err := trace.Emit(Event{
			Domain:    "test",
			Category:  "severity",
			Name:      sev.String(),
			Severity:  sev,
			FieldName: "level",
			Value:     sev.String(),
		}); err != nil {
			t.Fatalf("Emit failed for %s: %v", sev, err)
		}
	}

	if err := trace.Close(); err != nil {
		t.Fatalf("Close failed: %v", err)
	}

	files, err := filepath.Glob(filepath.Join(tmpDir, "*.dtj"))
	if err != nil {
		t.Fatalf("failed to glob dtj files: %v", err)
	}
	if len(files) == 0 {
		t.Fatal("no .dtj file created")
	}
}

// TestE2EIdempotentClose tests that Close is idempotent
func TestE2EIdempotentClose(t *testing.T) {
	if os.Getenv("DTJ_RUN_AGENT_E2E") != "1" {
		t.Skip("E2E test skipped. Set DTJ_RUN_AGENT_E2E=1 to run.")
	}

	tmpDir := t.TempDir()

	trace := Open(Config{
		ProducerName:    "e2e-test",
		ProducerVersion: "0.1.0",
		DataDir:         tmpDir,
	})
	if trace == nil {
		t.Fatal("Open returned nil")
	}

	if err := trace.Emit(Event{
		Domain:    "test",
		Category:  "cat",
		Name:      "event",
		Severity:  SeverityInfo,
		FieldName: "field",
		Value:     1,
	}); err != nil {
		t.Fatalf("Emit failed: %v", err)
	}

	// Close twice
	if err := trace.Close(); err != nil {
		t.Fatalf("Close failed: %v", err)
	}
	if err := trace.Close(); err != nil {
		t.Fatalf("Close second call failed: %v", err)
	}
}

File: packages/dtj-typescript/test/e2e.test.ts
/**
 * Opt-in E2E test for dtj-sdk.
 * Only runs when DTJ_RUN_AGENT_E2E=1 environment variable is set.
 * Requires a working dtj-agent binary and Unix socket support.
 */

import { test, describe, before, after } from "node:test";
import assert from "node:assert/strict";
import { TraceSession } from "../src/client.js";
import { existsSync, readFileSync, rmSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";

// Skip entire test suite if not opted in
const runE2E = process.env["DTJ_RUN_AGENT_E2E"] === "1";

if (!runE2E) {
  console.log("Skipping E2E tests. Set DTJ_RUN_AGENT_E2E=1 to run.");
  process.exit(0);
}

describe("E2E TraceSession", () => {
  let testDataDir: string;
  let sessionFile: string;

  before(() => {
    testDataDir = join(tmpdir(), `dtj-e2e-${Date.now()}`);
  });

  after(() => {
    try {
      rmSync(testDataDir, { recursive: true, force: true });
    } catch {}
  });

  test("creates trace session and emits event", async () => {
    const session = await TraceSession.open({
      producerName: "e2e-test",
      producerVersion: "0.1.0",
      dataDir: testDataDir,
      sessionFileName: "e2e-session.dtj",
    });

    // Should get a real TraceSession, not NoOpTraceSession
    assert.ok(session.constructor.name === "TraceSession");

    // Emit an event
    await session.emit({
      domain: "api",
      category: "request",
      name: "completed",
      severity: "info",
      fieldName: "duration_ms",
      value: 12.5,
      correlation: "request-42",
    });

    // Close session
    await session.close();

    // Verify .dtj file was created
    sessionFile = join(testDataDir, "e2e-session.dtj");
    assert.ok(existsSync(sessionFile), "Session file should exist");

    // Verify file has content
    const fileContent = readFileSync(sessionFile);
    assert.ok(fileContent.length > 0, "Session file should not be empty");

    // Verify it's a valid DTJ file by checking magic bytes
    // DTJ v1 magic: "DTJ\1" (0x44 0x54 0x4A 0x01)
    assert.equal(fileContent[0], 0x44); // 'D'
    assert.equal(fileContent[1], 0x54); // 'T'
    assert.equal(fileContent[2], 0x4A); // 'J'
    assert.equal(fileContent[3], 0x01); // version 1
  });

  test("emits multiple events with different value types", async () => {
    const session = await TraceSession.open({
      producerName: "e2e-test",
      producerVersion: "0.1.0",
      dataDir: testDataDir,
      sessionFileName: "e2e-multi.dtj",
    });

    // Emit boolean
    await session.emit({
      domain: "test",
      category: "bool",
      name: "flag",
      severity: "info",
      fieldName: "enabled",
      value: true,
    });

    // Emit integer
    await session.emit({
      domain: "test",
      category: "int",
      name: "count",
      severity: "info",
      fieldName: "value",
      value: 42,
    });

    // Emit bigint
    await session.emit({
      domain: "test",
      category: "bigint",
      name: "large",
      severity: "info",
      fieldName: "value",
      value: 123456789012345n,
    });

    // Emit float
    await session.emit({
      domain: "test",
      category: "float",
      name: "ratio",
      severity: "info",
      fieldName: "value",
      value: 3.14159,
    });

    // Emit bytes
    await session.emit({
      domain: "test",
      category: "bytes",
      name: "payload",
      severity: "info",
      fieldName: "data",
      value: new Uint8Array([0xDE, 0xAD, 0xBE, 0xEF]),
    });

    await session.close();

    const sessionFile = join(testDataDir, "e2e-multi.dtj");
    assert.ok(existsSync(sessionFile));
    const fileContent = readFileSync(sessionFile);
    assert.ok(fileContent.length > 0);
    assert.equal(fileContent[0], 0x44);
    assert.equal(fileContent[1], 0x54);
    assert.equal(fileContent[2], 0x4A);
    assert.equal(fileContent[3], 0x01);
  });

  test("handles different severity levels", async () => {
    const session = await TraceSession.open({
      producerName: "e2e-test",
      producerVersion: "0.1.0",
      dataDir: testDataDir,
      sessionFileName: "e2e-severity.dtj",
    });

    const severities = ["debug", "info", "warn", "error", "fatal"] as const;
    for (const sev of severities) {
      await session.emit({
        domain: "test",
        category: "severity",
        name: sev,
        severity: sev,
        fieldName: "level",
        value: sev,
      });
    }

    await session.close();

    const sessionFile = join(testDataDir, "e2e-severity.dtj");
    assert.ok(existsSync(sessionFile));
  });

  test("idempotent close", async () => {
    const session = await TraceSession.open({
      producerName: "e2e-test",
      producerVersion: "0.1.0",
      dataDir: testDataDir,
      sessionFileName: "e2e-close.dtj",
    });

    await session.close();
    await session.close(); // Should not throw
    assert.ok(true);
  });
});